Mods/SML/Source/SML/Private/Patching/BlueprintHookManager.cpp

```diff
@@ -84,53 +75,128 @@ void UBlueprintHookManager::InstallBlueprintHook(UFunction* Function, int32 Hook
+	//Begin writing function parameters - we have just the requested hook offset constant.
+	//The hook function needs this because it stores the user hooks according to their requested offset, not the resolved offset.
 	AppendedCode.Add(EX_IntConst);
-	WRITE_UNALIGNED(AppendedCode, int32, HookOffset);
+	WRITE_UNALIGNED(AppendedCode, int32, RequestedHookOffset);
 	AppendedCode.Add(EX_EndFunctionParms);
 
-
-	//Append original code that we replaced earlier with unconditional jump
-	AppendedCode.AddUninitialized(BytesAvailable);
-	FPlatformMemory::Memcpy(&AppendedCode[AppendedCode.Num() - BytesAvailable], &OriginalCode[HookOffset], BytesAvailable);
-
-	//Insert jump to original location for running code after hook
+	//Insert jump to after the resolved hook offset location for running code after hook
 	AppendedCode.Add(EX_Jump);
-	const int32 JumpDestination = HookOffset + BytesAvailable;
+	const int32 JumpDestination = ResolvedHookOffset + UBlueprintHookManager::JumpBytesRequired;
 	WRITE_UNALIGNED(AppendedCode, CodeSkipSizeType, JumpDestination);
 
-	//Finish generated code with EX_EndOfScript to avoid any surprises
-	AppendedCode.Add(EX_EndOfScript);
-
-
 	//Append generated code to the end of the function's original code now
 	const int32 StartOfAppendedCode = OriginalCode.Num();
 	OriginalCode.Append(AppendedCode);
 
-	//Fill space with EX_EndOfScript before replacement for safety
-	FPlatformMemory::Memset(&OriginalCode[HookOffset], EX_EndOfScript, BytesAvailable);
-
 	//Actually insert jump to the start of appended code to original hook location
-	OriginalCode[HookOffset] = EX_Jump;
-	FPlatformMemory::WriteUnaligned<CodeSkipSizeType>(&OriginalCode[HookOffset + 1], StartOfAppendedCode);
+	OriginalCode[ResolvedHookOffset] = EX_Jump;
+	FPlatformMemory::WriteUnaligned<CodeSkipSizeType>(&OriginalCode[ResolvedHookOffset + 1], StartOfAppendedCode);
 
 #if DEBUG_BLUEPRINT_HOOKING
-	DebugDumpFunctionScriptCode(Function, HookOffset, TEXT("AfterHook"));
+	DebugDumpFunctionScriptCode(Function, RequestedHookOffset, ResolvedHookOffset, TEXT("AfterHook"));
 #endif
 }
 
-int32 UBlueprintHookManager::PreProcessHookOffset(UFunction* Function, int32 HookOffset) {
-	if (HookOffset == EPredefinedHookOffset::Return) {
-		//For now Kismet Compiler will always generate only one Return node, so all
-		//execution paths will end up either with executing it directly or jumping to it
-		//So we need to hook only in one place to handle all possible execution paths
-		FSMLKismetBytecodeDisassembler Disassembler;
-		int32 ReturnOffset;
-		const bool bIsValid = Disassembler.FindFirstStatementOfType(Function, 0, EX_Return, ReturnOffset);
-		checkf(bIsValid, TEXT("EX_Return not found for function %s"), *Function->GetPathName());
-		return ReturnOffset;
+void UBlueprintHookManager::ModifyJumpTargetsForNewHookOffset(TArray<uint8>& Script, TSharedPtr<FJsonObject> Expression, int32 HookOffset)
+{
+	double OpcodeDouble;
```

Comment:
Why use a double here?

Reply:
Uh, because somehow I missed that there were a ton of overloads of this function. Fixed :)

Reply:
Ack